Add ERC: Transparent Financial Statements (#659)
ethereum/ERCs: ERCS/erc-7776.md

---
eip: 7776
title: Transparent Financial Statements

> description: A.P.I. for Protocols to Ensure Open Transparent Financial Data for all Investors.

Reviewer: First, we should cut out anything that is common to most ERCs:

```suggestion
description: Open Transparent Financial Data for all Investors.
```

Next, we need to add more details that aren't already in your title. You get about 140 characters for your description, so try to expand on ideas introduced in the proposal's title, keeping in mind that a good chunk of ERC readers come from a technical background (and not necessarily financial.) For example, what is open/transparent financial data?

Author: You are right. I changed it to:
> Open Transparent Financial Data such as Revenue, Costs of Goods Sold, Operating Expenses, EBITDA and Earnings Per Share for all Investors.

author: Ignacio Ceaglio (@Nachoxt17) <ignacioceaglio@gmail.com>
discussions-to: https://ethereum-magicians.org/t/erc-xxxx-transparent-financial-statements/21191
status: Review
type: Standards Track
category: ERC
created: 2024-09-20
requires: 20
---

## Abstract

> This proposal defines a standard A.P.I. that enables E.V.M. Blockchain-based companies (or also called "Protocols") to publish their financial information, specifically Income Statements and Balance Sheets, on-chain in a transparent and accessible manner through Solidity Smart Contracts. This standard aims to emulate the reporting structure used by publicly traded companies in traditional stocks markets, like the S.E.C. 10-Q filings. The financial statements include key information, namely as Revenue, Cost of Goods Sold, Operating Expenses, Operating Income, Earnings before Interest, Taxes, Depreciation, and Amortization (E.B.I.T.D.A.) and
> Earnings Per Token (E.P.Share-Token), allowing investors to assess the financial health of blockchain-based companies in a standardized, transparent, clear and reliable format.

Reviewer: The usual style for ERCs is to not use periods between letters in an initialism. Please make this change throughout the document. I won't be commenting on each line.

```suggestion
This proposal defines a standard API that enables EVM Blockchain-based companies (or also called "Protocols") to publish their financial information, specifically Income Statements and Balance Sheets, on-chain in a transparent and accessible manner through Solidity Smart Contracts. This standard aims to emulate the reporting structure used by publicly traded companies in traditional stocks markets, like the SEC 10-Q filings. The financial statements include key information, namely as Revenue, Cost of Goods Sold, Operating Expenses, Operating Income, Earnings before Interest, Taxes, Depreciation, and Amortization (EBITDA) and
Earnings Per Token (E.P.Share-Token), allowing investors to assess the financial health of blockchain-based companies in a standardized, transparent, clear and reliable format.
```

Author: Sure👍

## Motivation

> The motivation of this E.I.P. is to Bring Seriousness to the CryptoCurrencies Investments Market. Currently, the situation is as follows:

Reviewer: Weird capitalization, and probably a touch too informal.

```suggestion
The motivation of this EIP is to bring seriousness to the cryptocurrencies investments market. Currently, the situation is as follows:
```

Author: Okay👍

Most [ERC-20](./eip-20.md) Tokens representing E.V.M. Blockchain-based companies (or also called "Protocols"), DO NOT work the same way as a Publicly Traded Stock that represents a Share of ownership of the equity of that such company (so the user who buys a Protocol's ERC-20, is also now a share-holder and co-owner of the business, its profits and/or its dividends), but rather function as "Commodities" such as oil; they are consumable items created by said E.V.M. Blockchain-based company (or "Protocol") to be spent in their platform. They are Publicly Traded and advertised to be representing the underlying Protocol like a Share, working in practice the same way as a Commodity and without any Public, Transparent and _Clear_ Financial Information as publicly traded stocks have.

Added to that, most token research analysis reports that can be currently found on the internet are informal Substack or Twitter posts, with lots of abstract explanations about the features of the said Protocol to invest in, that lack of transparent financial numbers and factual financial information, that are made by anonymous users without real exposed reputations to affect.

> This E.I.P. will improve that by giving users and investors transparent, clear and factual financial information to work with when analyzing as a potential investment the such

Reviewer: I haven't read the specification section yet, but I suspect "factual" here doesn't mean "enforced by logic running on the blockchain." I'd be concerned you're overstating what this proposal can accomplish.

Author: Good point. In brief, the information is provided and upgraded by the same company/protocol that is using their own implementation of ERC-7776, directly from their business smart contracts (directly from their blockchain “bank accounts”).

Each protocol has different smart contracts and ways to generate revenue, name methods, they handle different assets, etc. So the Standard is made in a way to make that adjustable from different cases.

Why would they do that at all? Easy; it’s in their own interest to provide this information to generate trust among investors, and thus to potentially receive more investments.

I just added this line to the specification:

> If the contract owner uses data or methods from other owned smart contracts external to their smart contract implementation of this standard, those smart contracts MUST be verified in the correspondent blockchain explorer and of open and visible source code.

@SamWilsn  Do you think that there is any better way to enforce or explain this? The blockchain data already exists, my objective is that it is displayed in an open, clear and standardized way directly from the smart contracts of the protocol.

> E.V.M. Blockchain-based company that implements this E.I.P. in their Solidity Smart Contracts, and that will generate Trust, Transparency and Seriousness in the CryptoCurrencies Investments Market long term.

Reviewer: Weird capitalization again.

Author: Solved👍

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> All Transparent Financial Statements Standard implementations MUST implement ERC-20 to represent shares, and the financial numbers displayed in a StableCoin's value.

Reviewer: > the financial numbers displayed in a StableCoin's value

This is a bit vague. Could you elaborate?

Author: Yes:
> All Transparent Financial Statements Standard implementations MUST implement ERC-20 to represent shares, and the financial numbers such as Revenue, Costs of Goods Sold, Operating Expenses, Operating 
> Income, EBITDA, Other Income and Expenses, Net Income and Earnings Per Share MUST be displayed in the value of the protocol's stablecoin of choice.

All Transparent Financial Statements MUST implement ERC-20's optional metadata extensions.
The `name` and `symbol` functions SHOULD reflect the underlying token's `name` and `symbol` in some way.

> All methods MUST be of visibility `public`.

Reviewer: Why not external?

Author: Good question🤔. To be honest, I never thought about it. But since this information is meant to be consumed mainly by the public, it makes a little more sense for it to be `external`. Changed.

> All methods MUST return the `stablecoin`.

Reviewer: This is way too unclear. How would `supportsInterface` (for example) return the stable coin, or similarly, how can `ownerOf` return a stable coin?

Author: It was my definition mistake, I changed it:
> All methods MUST return their financial numbers valued in the provided `stablecoin`.

> _Timestamp Constraint_: For all methods, startTimestamp MUST be less than or equal to endTimestamp. If startTimestamp is equal to endTimestamp, the method returns a balance sheet snapshot. If startTimestamp is less than endTimestamp, the method returns an income statement for that period.

Reviewer: ```suggestion
_Timestamp Constraint_: For all methods, `startTimestamp` MUST be less than or equal to `endTimestamp`. If `startTimestamp` is equal to `endTimestamp`, the method returns a balance sheet snapshot. If `startTimestamp` is less than `endTimestamp`, the method returns an income statement for that period.
```

> _Output Structs_: Instead of a single `uint256` value, each method returns a `Struct` with one or _OPTIONAL_ more `uint256` entries to allow for detailed financial data, each one with their own customized entry `name`.

Reviewer: > `Struct`

Is `Struct` a type defined somewhere later, or do you mean `struct`?

Author: I meant `struct`. Changed.

### Definitions:

- Currency: The individual StableCoin used to value the publicly displayed financial numbers.
- Revenue: Total earnings from selling products or services before expenses.
- Cost of Goods Sold (C.O.G.S.): Direct costs for producing goods/services, including labor and materials.
- Operating Expenses: Expenses like Selling, General, and Administrative, Research and Development, and other operational costs.
- Operating Income: Revenue minus operating expenses.
- E.B.I.T.D.A.: Earnings Before Interest, Taxes, Depreciation, and Amortization.
- Other Income and Expenses: Non-operating income, such as interest, investment gains or losses.
- Net Income: Profit after all expenses, taxes, and deductions.
- E.P.S.: Earnings per Share Token (ERC-20), showing profit allocated per share.

### Methods

> #### currency

Reviewer: When referring to a method name, we prefer them to be enclosed in backticks (\`)

Author: No problem. Changed.

Returns the `address` of the individual StableCoin used to value the publicly displayed financial numbers.

```yaml

> - name: address

Reviewer: Is `address` correct here?

Author: Good point. I corrected it and improved it:
#### `stablecoinAddress`

> Returns the `address` of the individual stablecoin used to value the publicly displayed financial numbers.
> 
> ```functionyaml
> - name: stablecoinAddress
>   type: function

```yaml
- name: fullFinancialReport
  type: function
  stateMutability: view
  visibility: public
  inputs:
    - name: startTimestamp
      type: uint256
    - name: endTimestamp
      type: uint256
  outputs:
    - name: FullFinancialsStruct
      type: struct
      fields:
        - name: RevenueStruct
          type: struct
        - name: COGSStruct
          type: struct
        - name: OperatingExpensesStruct
          type: struct
        - name: OperatingIncomeStruct
          type: struct
        - name: EBITDAstruct
          type: struct
        - name: OtherIncomeExpensesStruct
          type: struct
        - name: NetIncomeStruct
          type: struct
        - name: EPSstruct
          type: struct

```

## Rationale

The current state of token investment analysis is opaque, with most information presented in an abstract and non-quantitative form. This standard A.P.I. ensures a consistent and reliable way for investors to evaluate blockchain projects based on real financial data published directly on-chain, not just speculative promises. This will establish greater

> trust in the cryptocurrency markets and align token analysis with the standards of traditional equity markets.

Reviewer: This paragraph probably belongs in Motivation because it's justifying the proposal as a whole. The Rationale section should explain choices made within this document, while the Motivation should convince the reader that the document itself is necessary (and a better choice than any competing proposals.) My favourite analogy is:

> Motivation: We need to build a shed because...
> Rationale: We decided to paint the shed red because...

Author: Good point. Understood. Changed.

ERC-20 is enforced because implementation details like Earnings Per Token calculation directly carry over to the accounting. This standardization makes the Transparent Financial Statements compatible with all ERC-20 use cases.

This implementation enables the Protocol to share their financial information both as their latest updated Balance Sheet (if the user chooses to just see a current snapshot of
the financial state of the company) and as an Income Statement (if the user chooses to see the evolution of the financial state of the company between two different block
timestamps) and also is thought to interact with other separated Smart Contracts of the same Protocol from which the financial information will be sent.

## Backwards Compatibility

Transparent Financial Statements Standard is fully backward compatible with the ERC-20 standard and has no known compatibility issues with other standards.

## Reference Implementation

> _NOTE: This Reference Implementation is a placeholder. It will be improved in the future from the feedback received._

Reviewer: ```suggestion
_NOTE: This Reference Implementation is a placeholder. It will be improved in the future from the feedback received._ <!-- TODO -->
```

If you use HTML-style comments, the linter will make sure you deal with them before you move to review. Otherwise it's way too easy to forget to replace these kind of notes.

Author: Good point. I will remove all of those `<!-- TODO -->`s when I get no more feedback for improvement.

```solidity

> // SPDX-License-Identifier: MIT

Reviewer: Code within the proposal itself must be `CC0-1.0`. If you'd like to keep this MIT, please move it to the `assets/` directory and link to it here.

Author: Okay. Corrected.

    }

    function fullFinancialReport(uint256 startTimestamp, uint256 endTimestamp) 
        public 
        view 
        returns (FullFinancialsStruct memory) 
    {
        require(startTimestamp <= endTimestamp, "Invalid timestamps");
        // Logic to calculate and return all financial details
        return FullFinancialsStruct(
            revenue(startTimestamp, endTimestamp),
            cogs(startTimestamp, endTimestamp),
            operatingExpenses(startTimestamp, endTimestamp),
            operatingIncome(startTimestamp, endTimestamp),
            ebitda(startTimestamp, endTimestamp),
            otherIncomeExpenses(startTimestamp, endTimestamp),
            netIncome(startTimestamp, endTimestamp),
            earningsPerShare(startTimestamp, endTimestamp)
        );
    }
}

```

## Security Considerations

This E.I.P. involves displaying critical financial data on-chain, so special attention must be paid to ensure the accuracy and security of the data, particularly in preventing tampering or manipulation of key financial figures. Further discussion on validation mechanisms and audits for the smart contracts containing financial data is necessary.

> Needs discussion.

Reviewer: ```suggestion
Needs discussion. <!-- TODO -->
```

All Transparent Financial Statements MUST implement ERC-20's optional metadata extensions.
The `name` and `symbol` functions SHOULD reflect the underlying token's `name` and `symbol` in some way.

All methods MUST be of visibility `external`.

All methods MUST return their financial numbers valued in the provided `stablecoin`.

If the contract owner uses data or methods from other owned smart contracts external to their smart contract implementation of this standard, those smart contracts MUST be verified in the correspondent 
blockchain explorer and of open and visible source code.

_Timestamp Constraint_: For all methods, `startTimestamp` MUST be less than or equal to `endTimestamp`. If `startTimestamp` is equal to `endTimestamp`, the method returns a balance sheet snapshot. If 
`startTimestamp` is less than `endTimestamp`, the method returns an income statement for that period.

_Output Structs_: Instead of a single `uint256` value, each method returns a `struct` with one or _OPTIONAL_ more `uint256` entries to allow for detailed financial data, each one with their own customized 
entry `name`.

> ### Definitions:

Reviewer: ```suggestion
### Definitions
```

- Currency: The individual stablecoin used to value the publicly displayed financial numbers.
- Revenue: Total earnings from selling products or services before expenses.
- Cost of Goods Sold (COGS): Direct costs for producing goods/services, including labor and materials.
- Operating Expenses: Expenses like Selling, General, and Administrative, Research and Development, and other operational costs.
- Operating Income: Revenue minus operating expenses.
- EBITDA: Earnings Before Interest, Taxes, Depreciation, and Amortization.
- Other Income and Expenses: Non-operating income, such as interest, investment gains or losses.
- Net Income: Profit after all expenses, taxes, and deductions.
- EPS: Earnings per Share Token (ERC-20), showing profit allocated per share.

### Methods

#### `stablecoinAddress`

Returns the `address` of the individual stablecoin used to value the publicly displayed financial numbers.

```yaml
- name: stablecoinAddress
  type: function
  visibility: external
  stateMutability: view

>   inputs:
>     /none/

Reviewer: ```suggestion
```

I think you can just omit inputs if there are none, but I am far from an expert on this YAML format.
